Sources/Workspace/InitPackage.swift

```diff
@@ -49,80 +30,227 @@ public final class InitPackage {
+    private func makePackageNew() throws {
+        let templateHomeDirectory = configPath.appending(components: "templates", "new-package")
+        var templateName: String?
+        
+        if let template = packageTemplateName {
+            guard fileSystem.exists(templateHomeDirectory.appending(component: template)) else {
+                throw InternalError("Could not find template folder: \(templateHomeDirectory.appending(component: template))")
+            }
+            
+            templateName = template
+        } else {
+            // Checking if a default template is present
+            if fileSystem.exists(templateHomeDirectory.appending(component: "default")) {
+                templateName = "default"
+                // There is a guard preventing '--type' to be used in conjunction with '--template'
+                // In the event that the defualt template is present and '--type' was used we'll infrom
+                // the user that the package type is coming from the template and not their supplied type.
+                if suppliedPackageType {
+                    progressReporter?("Package type is defined by the template 'default'.")
+                }
+            }
+        }
+        
+        if let template = templateName {
+            try fileSystem.getDirectoryContents(templateHomeDirectory.appending(component: template)).forEach {
+                progressReporter?("Copying \($0)")
+                try copyTemplate(fileSystem: fileSystem,
+                                 sourcePath: templateHomeDirectory.appending(components: template, $0),
+                                 destinationPath: destinationPath,
+                                 name: packageName)
+            }
+        } else {
+            let packageTemplate = PackageTemplate(sourcesDirectory: RelativePath("./Sources"),
+                                                  testsDirectory: nil,
+                                                  createSubDirectoryForModule: false,
+                                                  packageType: packageType)
+           
+            try writePackageStructure(template: packageTemplate)
+        }
+    }
+    
+    private func makePackageLegacy() throws {
+        let template = PackageTemplate(sourcesDirectory: RelativePath("./Sources"),
+                                       testsDirectory: RelativePath("./Tests"),
+                                       createSubDirectoryForModule: true,
+                                       packageType: packageType)
+        
+        try writePackageStructure(template: template)
+    }
+
+    private func copyTemplate(fileSystem: FileSystem, sourcePath: AbsolutePath, destinationPath: AbsolutePath, name: String) throws {
+        // Recursively copy the template package
+        // Currently only replaces the string literal "___NAME___", and "___NAME_AS_C99___"
+        let replaceName = "___NAME___"
+        let replaceNameC99 = "___NAME_AS_C99___"
```

Comment:
Nice — I think the templates just need to be updated to use this instead, then (I'd suggested a `___NAME_AS_MODULE___` I think).